### PR TITLE
Dev Docs - Overview and Types

### PR DESCRIPTION
**Pull request by jrevels**
This PR provides some much-needed developer documentation.

I've currently addressed the basics of the ForwardDiffNumber types, and added a small section for explaining why Julia is useful for FAD (@mlubin will probably recognize this).

~~I still want to explain the basics of how the API works, and the caching layer. After covering those two topics, I think this will be ready to merge.~~ I'm going to merge this as-is, and cover further topics in new PRs.

### CODE REVIEW

File: docs/source/types.rst [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+    i \leq j \leq N
+
+    i \leq k \leq j
+
+    M = \frac{N(N+1)(N+2)}{6}
+
+    d_p \in \{d_1, d_2, d_3...d_{M-1}, d_M\}
+
+    p_{ijkN} &= k + \left[\sum_{\alpha=i}^{j}\alpha-i+1\right] + \left[\sum_{\alpha=1}^{i-1} \sum_{\beta=\alpha}^{N}\beta-\alpha+1\right] \\
```
Comment (jrevels) [inline docs/source/types.rst]: This line appears to be incorrect. It's supposed to be formula for going from (i,j,k) --> p in the loop

```julia
function loop(N)
    p = 1
    for i in 1:N
        for j in i:N
            for k in i:j
                println((i,j,k), "-->", p)
                p += 1
            end
        end
    end
end
```

Must've made a math error somewhere...

**Comment by KristofferC** [comment]
Would it be possible to give some advice to what someone who wants to start hacking on this should focus on. Any performance bottle necks that would be good to look at our possible new features?  I always had the experience that just a few lines of what to look into can greatly reduce the barrier for a first commit.

**Comment by jrevels** [comment]
> Would it be possible to give some advice to what someone who wants to start hacking on this should focus on.

I think having a "How to Contribute" section is a good idea, thanks for the suggestion. Specifically, there's a lot of low-hanging fruit for anybody willing to implement new functions on ForwardDiffNumbers, or hand-optimize some of the auto-generated functions. I'll be sure to include a short tutorial for how to do so.

File: docs/source/types.rst [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+.. _`hyper-dual numbers`: https://adl.stanford.edu/hyperdual/Fike_AIAA-2011-886.pdf
+
+``Partials{T,C}`` 
+-----------------
+
+.. code-block:: julia
+    
+    immutable Partials{T,C}
+        data::C
+        Partials{N}(data::NTuple{N,T}) = new(data)
```
Comment (mlubin) [inline docs/source/types.rst]: Why do these need to be inner constructors?

**Comment by mlubin** [comment]
Did a quick read of the text and LGTM

File: docs/source/types.rst (hunk 2 above)
Comment (jrevels) [inline docs/source/types.rst]: To enforce `T == eltype(C)`. It also allows gives newcomers to the code an idea of what `C` is actually supposed to be. I could've used outer constructors here, I guess, but then the `T == eltype(C)` restriction wouldn't be inherent to the type definition.

**Pull request merged**